### DISCUSSION TITLE
Websocket Redirect Middleware

### DISCUSSION THREAD

**Discussion by x-stride** [fastapi/fastapi#9609] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import uvicorn
from fastapi import FastAPI, WebSocket
from fastapi.responses import RedirectResponse
from fastapi.datastructures import URL
from starlette.types import ASGIApp, Receive, Scope, Send


class WSRedirectMiddleware:
    def __init__(
            self,
            app: ASGIApp,
            redirect_paths: list
    ) -> None:
        self.app = app
        self.redirect_paths = redirect_paths

    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
        url = URL(scope=scope)
        if scope["type"] == "websocket" and url.path in self.redirect_paths:
            print(f'Got path: {url.path}')

            url = f'{scope["scheme"]}://{url.hostname}:{url.port}/ws2'
            print(f'Redirect to: {url}')

            response = RedirectResponse(url, status_code=307)
            await response(scope, receive, send)
        else:
            await self.app(scope, receive, send)


application = FastAPI()


# Redirect from /ws1 to /ws2
application.add_middleware(WSRedirectMiddleware, redirect_paths=['/ws1'])


@application.websocket("/ws1")
async def websocket(ws: WebSocket):
    await ws.accept()
    await ws.send_json({"msg": "Hello WebSocket1"})
    await ws.close()


@application.websocket("/ws2")
async def websocket(ws: WebSocket):
    await ws.accept()
    await ws.send_json({"msg": "Hello WebSocket2"})
    await ws.close()

if __name__ == '__main__':
    uvicorn.run(application, host="localhost", port=8088)
```

### Description

Hello all,

Created an alternative of [HTTPSRedirectMiddleware](https://github.com/encode/starlette/blob/master/starlette/middleware/httpsredirect.py) for redirection, but it's not working as I expected.

**Error logs:**
```
INFO:     Started server process [25196]
INFO:     Waiting for application startup.
INFO:     ASGI 'lifespan' protocol appears unsupported.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://localhost:8088 (Press CTRL+C to quit)
Got path: /ws1
Redirect to: ws://localhost:8088/ws2
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "C:\Users\user\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\uvicorn\protocols\websockets\websockets_impl.py", line 225, in run_asgi
    result = await self.app(self.scope, self.asgi_receive, self.asgi_send)
  File "C:\Users\user\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\uvicorn\middleware\proxy_headers.py", line 78, in __call__
    return await self.app(scope, receive, send)
  File "C:\Users\user\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\fastapi\applications.py", line 270, in __call__
    await super().__call__(scope, receive, send)
  File "C:\Users\user\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\starlette\applications.py", line 124, in __call__
    await self.middleware_stack(scope, receive, send)
  File "C:\Users\user\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\starlette\middleware\errors.py", line 149, in __call__
    await self.app(scope, receive, send)
  File "C:\Users\user\Documents\Signalling\main.py", line 24, in __call__
    await response(scope, receive, send)
  File "C:\Users\user\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\starlette\responses.py", line 160, in __call__
    await send(
  File "C:\Users\user\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\uvicorn\protocols\websockets\websockets_impl.py", line 288, in asgi_send
    raise RuntimeError(msg % message_type)
RuntimeError: Expected ASGI message 'websocket.accept' or 'websocket.close', but got 'http.response.start'.
INFO:     connection open
INFO:     connection closed
```

Checking the FastAPI [httpsredirect.py](https://github.com/tiangolo/fastapi/blob/master/fastapi/middleware/httpsredirect.py) and various middleware-guides out there I thought this would work... but looking at the [ASGI spec](https://asgi.readthedocs.io/en/latest/specs/www.html) it seems impossible...

**Or can it be done?**

With best regards,

/x-stride

### Operating System

Windows

### FastAPI Version

0.85.0

### Python Version

3.10.11

**Comment by MRigal** [comment]
@tiangolo I would also be interested in that. My use case is the need to have an application with FastAPI serving the data and a Panel Server rendering graphs via Websockets, running on a different port. But I can expose only one port to the web and thus I would like a WebsocketRedirect from one route to the Panel Server.

**Reply by WSH032** [reply]
Hi, can [fastapi-proxy-lib](https://github.com/WSH032/fastapi-proxy-lib/) solve your problem? I specifically designed it to address this issue.

```py
from fastapi_proxy_lib.fastapi.app import reverse_ws_app
from httpx import AsyncClient

client = AsyncClient()
base_url = "ws://echo.websocket.events/"

app = reverse_ws_app(client=client, base_url=base_url)
```

---

Or, may be you will prefer using a `router` instead of an `app`:

<https://wsh032.github.io/fastapi-proxy-lib/reference/fastapi_proxy_lib/fastapi/router/#fastapi_proxy_lib.fastapi.router.RouterHelper--examples>

**Reply by MRigal** [reply]
Thanks for the library, I could handle my problem in between. You may also advertise it here: https://github.com/encode/starlette/discussions/2167

**Reply by WSH032** [reply]
It seems like a different issue. `fastapi-proxy-lib` implements forwarding proxy, not redirection.

Forwarding proxy might be better suited for your requirements rather than redirection, so I advertise it for you. (forgive my audacity.)

However, `fastapi-proxy-lib` does not implement the feature of redirection.

---

To address this issue, I think redirecting the HTTP handshake before establishing the WebSocket connection should suffice.

**Reply by x-stride** [reply]
A bit late in the reply, but yes, resolved :)

**Comment by Kludex** [comment, marked as answer]
I think the `HTTPSRedirectMiddleware` redirect works for WebSockets now.
